timemachine/md/smc.py

```diff
@@ -120,6 +122,158 @@ def accumulate_results(samples, indices, log_weights, incremental_log_weights):
+       starting at lambda=0.
+
+    Parameters
+    ----------
+    samples: [N,] list
+    propagate: function
+        [move(x, lam) for x in xs]
+        for example, move(x, lam) might mean "run 100 steps of all-atom MD targeting exp(-u(., lam)), initialized at x"
+    log_prob: function
+        [exp(-u(x, lam, first: bool)) for x in xs]
+        first is set to True for the first iteration of each binary search.
+        This may be used to improve performance by caching prefactors.
+    resample: function
+        (optionally) perform resampling given an array of log weights
+    cess_target: float
+        Target CESS (see `conditional_effective_sample_size`). Intermediate lambdas
+        will be sampled keeping the CESS between successive windows at approximately
+        this value. This value should be in the range (1, N).
+    epsilon:
+        Used to determine the precision of the adaptive binary search.
+    store_intermediate_traj:
+        Set to True (default) to store intermediate trajectories.
+
+    Returns
+    -------
+    trajs_dict
+        "traj"
+            [K-1, N] list of snapshots only if `store_intermediate_traj` = True.
+            [1, N] list of snapshots if `store_intermediate_traj` = False.
+        "incremental_log_weights_traj"
+            [K-1, N] array of incremental log weights
+        "ancestry_traj"
+            [K-1, N] array of ancestor idxs
+        "log_weights_traj"
+            [K, N] array of accumulated log weights
+        "lambdas_traj"
+            [K] array of adaptive lambdas
+
+    References
+    ----------
+    * [Zhou, Johansen, Aston, 2016]
+        Towards Automatic Model Comparison: An Adaptive Sequential Monte Carlo Approach
+        https://arxiv.org/pdf/1303.3123 (Algorithm #4)
+    """
+    n = len(samples)
+
+    # check cess_target
+    assert cess_target > 1, f"cess_target is too small: {cess_target} <= 1"
+    assert cess_target < n, f"cess_target is too large: {cess_target} >= {n}"
+
+    log_weights = np.zeros(n)
+    norm_log_weights = log_weights - logsumexp(log_weights)
+
+    # store
+    sample_traj = [samples]
+    ancestry_traj = [np.arange(n)]
+    log_weights_traj = [np.array(log_weights)]
+    incremental_log_weights_traj = []  # note: redundant but convenient
+    lambdas_traj = [0.0]
+
+    def accumulate_results(samples, indices, log_weights, incremental_log_weights, lam_target):
+        if store_intermediate_traj:
+            sample_traj.append(samples)
+        else:
+            # only store one intermediate set of samples to reduce memory usage
+            sample_traj[0] = samples
+        ancestry_traj.append(indices)
+        log_weights_traj.append(np.array(log_weights))
+        incremental_log_weights_traj.append(np.array(incremental_log_weights))
+        lambdas_traj.append(lam_target)
+
+    lam_initial = 0.0
+    lam_target = 1.0  # adapted
+
+    # Main ASMC loop
+    while True:
```

Comment:
nit: possibly replace with a for loop up to some max iteration count?

Reply:
I'm not sure how that would work. If we hit the max count what do we return? Just the partial results?

Reply:
hmm, was thinking either:
* after max iteration count limit is reached, jump directly to lam=1
* raise an error

unsure if either is preferable to potentially unbounded loop

another approach to ensure the loop is bounded is to set a minimum step size epsilon, so each binary search looks at the interval `(lam_current + eps, lam_target]` rather than `(lam_current, lam_target)`

Reply:
This should be handled by xtol in the solver though. I tested this by setting cess_target=len(samples) and confirmed we don't get stuck in an infinite loop.

Reply:
>This should be handled by xtol in the solver though. I tested this by setting cess_target=len(samples) and confirmed we don't get stuck in an infinite loop.

Ahh, I missed that interaction.

(So this implicitly sets max iteration count = 1.0 / xtol, which by default is 100.)

Minor point, but controlling the solver precision and max # lambda windows separately seems preferable (binary search is cheap, esp. with cached prefactors, and a user may be tempted to make the solver very precise, e.g. setting xtol to machine epsilon or so. May want to be able to do this independent of max iteration count.

Reply:
Generally you can do this by setting a reasonable cess_target and set xtol to be small. It's only `max iteration count = 1.0 / xtol` when cess_target ~= n_samples.

Reply:
>It's only max iteration count = 1.0 / xtol when cess_target ~= n_samples.

Or, potentially, if something else unspecified goes wrong. (E.g. if all the MD simulations blow up, or if something else happens so that infinitesimally small changes in lambda result in very large changes in U(x, lam)...)

Reply:
good point.

Reply:
Added max iterations and raise if exceeded